l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,638 @@
+    /// @param _l2TxGasPerPubdataByte The gasPerPubdataByteLimit to be used in the corresponding L2 transaction
+    /// @param _refundRecipient The address on L2 that will receive the refund for the transaction.
+    /// @dev If the L2 deposit finalization transaction fails, the `_refundRecipient` will receive the `_l2Value`.
+    /// Please note, the contract may change the refund recipient's address to eliminate sending funds to addresses
+    /// out of control.
+    /// - If `_refundRecipient` is a contract on L1, the refund will be sent to the aliased `_refundRecipient`.
+    /// - If `_refundRecipient` is set to `address(0)` and the sender has NO deployed bytecode on L1, the refund will
+    /// be sent to the `msg.sender` address.
+    /// - If `_refundRecipient` is set to `address(0)` and the sender has deployed bytecode on L1, the refund will be
+    /// sent to the aliased `msg.sender` address.
+    /// @dev The address aliasing of L1 contracts as refund recipient on L2 is necessary to guarantee that the funds
+    /// are controllable through the Mailbox, since the Mailbox applies address aliasing to the from address for the
+    /// L2 tx if the L1 msg.sender is a contract. Without address aliasing for L1 contracts as refund recipients they
+    /// would not be able to make proper L2 tx requests through the Mailbox to use or withdraw the funds from L2, and
+    /// the funds would be lost.
+    /// @return l2TxHash The L2 transaction hash of deposit finalization.
+    function depositLegacyErc20Bridge(
+        address _prevMsgSender,
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _amount,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) external payable override onlyLegacyBridge nonReentrant returns (bytes32 l2TxHash) {
+        require(l2BridgeAddress[ERA_CHAIN_ID] != address(0), "ShB b. n dep");
+        require(_l1Token != l1WethAddress, "ShB: WETH deposit not supported 2");
+
+        // Note that funds have been transferred to this contract in the legacy ERC20 bridge.
+        if (!hyperbridgingEnabled[ERA_CHAIN_ID]) {
+            chainBalance[ERA_CHAIN_ID][_l1Token] += _amount;
+        }
+
+        bytes memory l2TxCalldata = _getDepositL2Calldata(_prevMsgSender, _l2Receiver, _l1Token, _amount);
+
+        {
+            // If the refund recipient is not specified, the refund will be sent to the sender of the transaction.
+            // Otherwise, the refund will be sent to the specified address.
+            // If the recipient is a contract on L1, the address alias will be applied.
+            address refundRecipient = _refundRecipient;
+            if (_refundRecipient == address(0)) {
+                refundRecipient = _prevMsgSender != tx.origin
+                    ? AddressAliasHelper.applyL1ToL2Alias(_prevMsgSender)
+                    : _prevMsgSender;
```

Comment:
This method of determining if the sender of the transaction is a smart contract or a human is discriminatory against someone using smart contract wallets such as 1/1 multi-sig safe wallets as opposed to someone using metamask.

Not a security concern. Just wanted people to be fully aware of the implications of this line of code.

Reply:
I'm guessing cross-chain attacks?

Reply:
@saxenism Right, or the exact example is:
- User deploys contract on L2 with address `x`
- `x` contract is immutable and other contracts relies on the code invariants by `require(msg.sender == x);`
- After some time deployed protocol became popular
- User deploys contract with the same address on L1 and sends transaction on behalf of `x`

So, auditors should not only check all L2 contracts but also all deployed and potentially deployed L1 contracts